fw/http_parser.c

```diff
@@ -6558,56 +6557,71 @@ __h2_req_parse_content_type(TfwHttpMsg *hm, unsigned char *data, size_t len,
 
 	__FSM_STATE(I_ContType) {
 		if (req->method != TFW_HTTP_METH_POST)
-			__FSM_I_JMP(I_EoL);
+			__FSM_H2_I_JMP(I_EoL);
 		/* Fall through. */
 	}
 
 	__FSM_STATE(I_ContTypeMediaType) {
-		static const TfwStr s_multipart_form_data =
-			TFW_STR_STRING("multipart/form-data");
-		H2_TRY_STR_LAMBDA_fixup(&s_multipart_form_data, &parser->hdr, {
-			__set_bit(TFW_HTTP_B_CT_MULTIPART, req->flags);
-			__FSM_EXIT(CSTR_EQ);
-		}, I_ContTypeMediaType, I_ContTypeMaybeMultipart);
-		if (chunk->len >= sizeof("multipart/") - 1) {
-			TRY_STR_INIT();
-			__FSM_I_JMP(I_ContTypeOtherSubtype);
-		} else {
-			TRY_STR_INIT();
-			__FSM_I_JMP(I_ContTypeOtherType);
-		}
+		static const TfwStr s_multipart_slash =
+			TFW_STR_STRING("multipart/");
+		H2_TRY_STR_2FIN_fixup(&s_multipart_slash, &parser->hdr, {
+				__FSM_EXIT(CSTR_NEQ);
+			}, {
+				__FSM_EXIT(CSTR_EQ);
+			}, I_ContTypeMediaType, I_ContTypeAfterMultipartSlash);
+		TRY_STR_INIT();
+		__FSM_H2_I_JMP(I_ContTypeOtherType);
+	}
+
+	__FSM_STATE(I_ContTypeAfterMultipartSlash) {
+		static const TfwStr s_form_data =
+			TFW_STR_STRING("form-data");
+		H2_TRY_STR_2FIN_fixup(&s_form_data, &parser->hdr, {
+				__set_bit(TFW_HTTP_B_CT_MULTIPART, req->flags);
+				__FSM_EXIT(CSTR_EQ);
+			}, {
+				__FSM_EXIT(CSTR_EQ);
+			}, I_ContTypeAfterMultipartSlash, I_ContTypeMaybeMultipart);
+		TRY_STR_INIT();
+		__FSM_H2_I_JMP(I_ContTypeOtherSubtype);
 	}
```

Comment:
Let's not to introduce the new state: frankly the code is still very sophisticated and the change is doubtful from the performance point of view. Let's just clarify the previous change about `p` movement in `H2_TRY_STR_3LAMBDA_fixup`.